functions/src/notifications.ts

```diff
@@ -92,6 +92,75 @@ export function createActivityNotification(token: string, activity_id: string):
     }
 }
 
+// Sends a message to a chat if a user leaves or joins it:
+export async function onActivityUsersChanged(change: Change<FirebaseFirestore.DocumentSnapshot>, context: EventContext) {
+    const activityId: string = change.after.id
+    const usersBefore = getUsersFromChange(change.before)
+    const usersAfter = getUsersFromChange(change.after)
+
+    if (usersBefore.length !== usersAfter.length) {
+        // Gets which user joined or left the activity. changedUser will be that person's UID.
+        const changedUser = getDifferentString(usersBefore, usersAfter)
+        if (! changedUser) return -1
+
+        // @ts-ignore because it doesn't like the mock
+        const userInfo = await Refs(database).user(changedUser).get()
+        if (userInfo.exists) {
+            const changedUserInfo = userInfo.data()
+            // Setup message based on whether they left or joined:
+            const msgBody = `${changedUserInfo.name} has ${usersBefore > usersAfter ? 'left' : 'joined'} your activity.`
+            
+            return exports.sendChatMessage(activityId, msgBody, changedUser, new Date())
+        }
+    }
+    return 0
+}
+
+// Sends a user left/joined message to the specified activity
+export function sendChatMessage(activityId: string, message: string, sender: string, date: Date) {
+    // @ts-ignore because it doesn't like the database mock:
+    const chatRef = Refs(database).chat(activityId)
+    return chatRef.add({
+        message,
+        sender,
+        date_sent: date,
+        type: 'user_join_or_leave'
```

Comment:
Just curious about this, maybe you could explain it to me some time. Does adding a message automatically trigger the notification in another cloud function?  That's pretty neat.

Reply:
Yep! Adding the message updates the firestore ref, which then calls the triggers (if any) on that ref. In our case, we're listening for notifications there so it's a nice win.

As the docs say:
>Note: Any time you write to the same document that triggered a function, you are at risk of creating an infinite loop. Use caution and ensure that you safely exit the function when no change is needed.